pkg/hubble/recorder/sink/dispatch.go

```diff
@@ -0,0 +1,256 @@
+}
+
+func (d *Dispatch) decodeRecordCaptureLocked(data []byte) (rec record, err error) {
+	dataLen := uint32(len(data))
+	if dataLen < monitor.RecorderCaptureLen {
+		return record{}, fmt.Errorf("not enough data to decode capture message: %d", dataLen)
+	}
+
+	// This needs to stay in sync with struct capture_msg from
+	// bpf/include/pcap.h.
+	// We could use binary.Read on monitor.RecorderCapture, but since it
+	// requires reflection, it is too slow to use on the critical path here.
+	const (
+		offsetRuleID         = 2
+		offsetTimeBoot       = 8
+		offsetCaptureLength  = 16
+		offsetOriginalLength = 20
+	)
+	n := byteorder.Native
+	ruleID := n.Uint16(data[offsetRuleID:])
+	timeBoot := n.Uint64(data[offsetTimeBoot:])
+	capLen := n.Uint32(data[offsetCaptureLength:])
+	origLen := n.Uint32(data[offsetOriginalLength:])
+
+	// data may contain trailing garbage from the perf ring buffer
+	// https://lore.kernel.org/patchwork/patch/1244339/
+	packetEnd := monitor.RecorderCaptureLen + capLen
+	if dataLen < packetEnd {
+		return record{}, fmt.Errorf("capture record too short: want:%d < got:%d", dataLen, packetEnd)
+	}
+	packet := data[monitor.RecorderCaptureLen:packetEnd]
+
+	return record{
+		timestamp: time.Unix(0, d.bootTimeOffset+int64(timeBoot)),
+		ruleID:    ruleID,
+		inclLen:   capLen,
+		origLen:   origLen,
+		data:      packet,
+	}, nil
+}
+
+const estimationRounds = 25
+
+func estimateBootTimeOffset() (bootTimeOffset int64, err error) {
+	// The datapath is currently using ktime_get_boot_ns for the pcap timestamp,
+	// which corresponds to CLOCK_BOOTTIME. To be able to convert the the
+	// CLOCK_BOOTTIME to CLOCK_REALTIME (i.e. a unix timestamp).
+
+	// There can be an arbitrary amount of time between the execution of
+	// time.Now() and unix.ClockGettime() below, especially under scheduler
+	// pressure during program startup. To reduce the error introduced by these
+	// delays, we pin the current Go routine to its OS thread and measure the
+	// clocks multiple times, taking only the smallest observed difference
+	// between the two values (which implies the smallest possible delay
+	// between the two snapshots).
+	var minDiff int64 = 1<<63 - 1
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread() // unlocking twice is a no-op
+	for round := 0; round < estimationRounds; round++ {
+		var bootTimespec unix.Timespec
+
+		// Ideally we would use __vdso_clock_gettime for both clocks here,
+		// to have as little overhead as possible.
+		// time.Now() will actually use VDSO on Go 1.9+, but calling
+		// unix.ClockGettime to obtain CLOCK_BOOTTIME is a regular system call
+		// for now.
+		unixTime := time.Now()
+		err = unix.ClockGettime(unix.CLOCK_BOOTTIME, &bootTimespec)
+		if err != nil {
+			return 0, err
+		}
+
+		offset := unixTime.UnixNano() - bootTimespec.Nano()
+		diff := offset
+		if diff < 0 {
+			diff = -diff
+		}
+
+		if diff < minDiff {
+			minDiff = diff
+			bootTimeOffset = offset
+		}
+	}
+	runtime.UnlockOSThread()
```

Comment:
Nit: this can be removed.

Reply:
Good catch, thanks! Originally I had more code before the return statement so I wanted to do an early unlock, but that's not needed anymore in the current version.

Will remove once CI is green (to avoid re-triggering it)